// tdesign-react/cjs/table/hooks/useTreeDataExpand.js
/**
 * tdesign v1.7.1
 * (c) 2024 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../_chunks/dep-ff9a5a06.js');
var toConsumableArray = require('../../_chunks/dep-987bfa39.js');
var slicedToArray = require('../../_chunks/dep-c86b4e2f.js');
var React = require('react');
var hooks_usePrevious = require('../../hooks/usePrevious.js');
var hooks_useControlled = require('../../hooks/useControlled.js');
var _common_js_table_treeStore = require('../../_common/js/table/tree-store.js');
require('../../_chunks/dep-2e4b81d6.js');
require('../../_chunks/dep-58dbd853.js');
require('../../_chunks/dep-c0841a72.js');
require('../../_chunks/dep-174c24e3.js');
require('../../_chunks/dep-04f46890.js');
require('../../_chunks/dep-a4fb1294.js');
require('../../_chunks/dep-68d6ca4f.js');
require('../../_chunks/dep-59101a1c.js');
require('../../_chunks/dep-e7a0e8b4.js');
require('../../_chunks/dep-f70c56e2.js');
require('../../_chunks/dep-315a4a2b.js');
require('../../_util/noop.js');
require('../../_chunks/dep-405223a6.js');
require('../../_chunks/dep-0f313135.js');
require('../../_chunks/dep-7c29b5a1.js');
require('../../_chunks/dep-55437cd3.js');
require('../../_chunks/dep-4a411278.js');
require('../../_chunks/dep-9171afa5.js');
require('../../_chunks/dep-7d38d5e1.js');
require('../../_chunks/dep-5ed3b3ae.js');
require('../../_chunks/dep-8bb91473.js');
require('../../_chunks/dep-9f572ddd.js');
require('../../_chunks/dep-2a42b204.js');
require('../../_chunks/dep-d93c04ca.js');
require('../../_chunks/dep-2a8c4619.js');
require('../../_chunks/dep-6bae679a.js');
require('../../_common/js/table/utils.js');
require('../../_chunks/dep-0bcf83a3.js');
require('../../_common/js/log/log.js');

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function useTreeDataExpand(props, params) {
  var store = params.store,
    dataSource = params.dataSource,
    rowDataKeys = params.rowDataKeys,
    setDataSource = params.setDataSource;
  var data = props.data,
    tree = props.tree;
  var _useState = React.useState(false),
    _useState2 = slicedToArray._slicedToArray(_useState, 2),
    isDefaultExpandAllExecute = _useState2[0],
    setIsDefaultExpandAllExecute = _useState2[1];
  var _useControlled = hooks_useControlled["default"](props, "expandedTreeNodes", props.onExpandedTreeNodesChange, {
      defaultExpandedTreeNodes: props.defaultSelectedRowKeys || []
    }),
    _useControlled2 = slicedToArray._slicedToArray(_useControlled, 2),
    tExpandedTreeNode = _useControlled2[0],
    setTExpandedTreeNode = _useControlled2[1];
  var oldExpandedTreeNode = hooks_usePrevious["default"](tExpandedTreeNode);
  var _useState3 = React.useState({
      type: "props-change"
    }),
    _useState4 = slicedToArray._slicedToArray(_useState3, 2),
    changedExpandTreeNode = _useState4[0],
    setChangedExpandTreeNode = _useState4[1];
  function expandAll() {
    var type = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : "expand-all";
    var list = arguments.length > 1 ? arguments[1] : undefined;
    var newData = list || data;
    setDataSource(store.expandAll(newData, rowDataKeys));
    var expandedNode = dataSource.map(function (t) {
      return _common_js_table_treeStore.getUniqueRowValue(t, rowDataKeys.rowKey);
    });
    setTExpandedTreeNode(expandedNode, {
      row: void 0,
      rowState: void 0,
      rowIndex: void 0,
      type: "expand",
      trigger: type
    });
    setChangedExpandTreeNode({
      type: "expand-all"
    });
  }
  function foldAll() {
    setDataSource(toConsumableArray._toConsumableArray(store.foldAll(dataSource, rowDataKeys)));
    setTExpandedTreeNode([], {
      row: void 0,
      rowState: void 0,
      rowIndex: void 0,
      type: "fold",
      trigger: "fold-all"
    });
  }
  function onExpandFoldIconClick(p, trigger) {
    var _props$onTreeExpandCh;
    var row = p.row,
      rowIndex = p.rowIndex;
    setChangedExpandTreeNode(_objectSpread({
      type: "user-reaction-change"
    }, p));
    var rowValue = _common_js_table_treeStore.getUniqueRowValue(row, rowDataKeys.rowKey);
    var rowState = store.treeDataMap.get(rowValue);
    var expandedNodes = toConsumableArray._toConsumableArray(tExpandedTreeNode);
    if (rowState.expanded) {
      var expandedChildrenKeys = store.getExpandedChildrenKeys([row], rowDataKeys);
      for (var i = 0, len = expandedNodes.length; i < len; i++) {
        var nodeValue = expandedNodes[i];
        if (expandedChildrenKeys.includes(nodeValue)) {
          expandedNodes[i] = void 0;
        }
      }
      expandedNodes = expandedNodes.filter(Boolean);
    } else {
      expandedNodes.push(rowValue);
    }
    var params2 = {
      row: row,
      rowIndex: rowIndex,
      rowState: rowState,
      trigger: trigger
    };
    setTExpandedTreeNode(expandedNodes, _objectSpread(_objectSpread({}, params2), {}, {
      type: rowState.expanded ? "fold" : "expand"
    }));
    (_props$onTreeExpandCh = props.onTreeExpandChange) === null || _props$onTreeExpandCh === void 0 || _props$onTreeExpandCh.call(props, params2);
  }
  function updateExpandState(data2, tExpandedTreeNode2) {
    var oldExpandedTreeNode2 = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : [];
    var _diffExpandedTreeNode = _common_js_table_treeStore.diffExpandedTreeNode(tExpandedTreeNode2, oldExpandedTreeNode2),
      addedList = _diffExpandedTreeNode.addedList,
      removedList = _diffExpandedTreeNode.removedList;
    store.expandTreeNode(addedList, data2, rowDataKeys);
    store.foldTreeNode(removedList, data2, rowDataKeys);
    return data2;
  }
  React.useEffect(function () {
    if (!store.treeDataMap.size) return;
    if (changedExpandTreeNode.type === "user-reaction-change") {
      var _ref = changedExpandTreeNode || {},
        row = _ref.row,
        rowIndex = _ref.rowIndex;
      var newData = store.toggleExpandData({
        row: row,
        rowIndex: rowIndex
      }, dataSource, rowDataKeys);
      setDataSource(toConsumableArray._toConsumableArray(newData));
    } else if (changedExpandTreeNode.type === "props-change") {
      updateExpandState(toConsumableArray._toConsumableArray(dataSource), tExpandedTreeNode, oldExpandedTreeNode);
    }
    if (changedExpandTreeNode.type !== "props-change") {
      setChangedExpandTreeNode({
        type: "props-change"
      });
    }
  }, [tExpandedTreeNode]);
  var updateExpandOnDataChange = function updateExpandOnDataChange(data2) {
    if (tree !== null && tree !== void 0 && tree.defaultExpandAll && !isDefaultExpandAllExecute) {
      expandAll("default-expand-all", toConsumableArray._toConsumableArray(data2));
      setIsDefaultExpandAllExecute(true);
    } else if (tExpandedTreeNode !== null && tExpandedTreeNode !== void 0 && tExpandedTreeNode.length) {
      var newData = updateExpandState(toConsumableArray._toConsumableArray(data2), tExpandedTreeNode, []);
      setDataSource(toConsumableArray._toConsumableArray(newData));
    }
  };
  return {
    tExpandedTreeNode: tExpandedTreeNode,
    isDefaultExpandAllExecute: isDefaultExpandAllExecute,
    expandAll: expandAll,
    foldAll: foldAll,
    onExpandFoldIconClick: onExpandFoldIconClick,
    updateExpandOnDataChange: updateExpandOnDataChange
  };
}

exports["default"] = useTreeDataExpand;
exports.useTreeDataExpand = useTreeDataExpand;
