// tdesign-react/cjs/_util/useDrag.js
/**
 * tdesign v1.7.1
 * (c) 2024 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');

var useDrag = function useDrag(ref, options) {
  var start = options.start,
    end = options.end,
    drag = options.drag;
  var isDraggingRef = React.useRef(false);
  var getCoordinate = function getCoordinate(event) {
    try {
      var _ref$current;
      var rect = (_ref$current = ref.current) === null || _ref$current === void 0 ? void 0 : _ref$current.getBoundingClientRect();
      var left = event.clientX - rect.left;
      var top = event.clientY - rect.top;
      return {
        y: Math.min(Math.max(0, top), rect.height),
        x: Math.min(Math.max(0, left), rect.width)
      };
    } catch (error) {
      return {
        y: null,
        x: null
      };
    }
  };
  var handlePointerMove = function handlePointerMove(e) {
    if (isDraggingRef.current) drag(getCoordinate(e), e);
  };
  var handlePointerUp = function handlePointerUp(e) {
    isDraggingRef.current = false;
    end(getCoordinate(e), e);
    var element = ref.current;
    element.removeEventListener("mouseup", handlePointerUp);
    element.removeEventListener("mousemove", handlePointerMove);
  };
  var handlePointerDown = function handlePointerDown(e) {
    isDraggingRef.current = true;
    start(getCoordinate(e), e);
    var element = ref.current;
    element.addEventListener("mouseup", handlePointerUp);
    element.addEventListener("mousemove", handlePointerMove);
  };
  React.useEffect(function () {
    var element = ref.current;
    if (element) {
      element.addEventListener("mousedown", handlePointerDown);
    }
    return function () {
      if (element) {
        element.removeEventListener("mousedown", handlePointerDown);
      }
    };
  }, []);
  return {
    isDragging: isDraggingRef.current
  };
};

exports["default"] = useDrag;
