// tdesign-react/esm/_common/js/tree-v1/tree-node-model.js
/**
 * tdesign v1.7.1
 * (c) 2024 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../../../_chunks/dep-9d06ffd4.js';
import { i as isUndefined_1 } from '../../../_chunks/dep-f2dd6554.js';
import { i as isBoolean_1 } from '../../../_chunks/dep-c248f47f.js';
import { p as pick_1 } from '../../../_chunks/dep-e62d6efe.js';
import { o as omit_1 } from '../../../_chunks/dep-b112394d.js';
import log from '../log/log.js';
import '../../../_chunks/dep-08fc9b22.js';
import '../../../_chunks/dep-f3bac8f4.js';
import '../../../_chunks/dep-9d4210ce.js';
import '../../../_chunks/dep-dc4499f2.js';
import '../../../_chunks/dep-49b9df8b.js';
import '../../../_chunks/dep-17c2aa2b.js';
import '../../../_chunks/dep-c3a67e8f.js';
import '../../../_chunks/dep-cc3b0635.js';
import '../../../_chunks/dep-39d65f32.js';
import '../../../_chunks/dep-70de8f55.js';
import '../../../_chunks/dep-c2c3c04d.js';
import '../../../_chunks/dep-48a2bd8e.js';
import '../../../_chunks/dep-8cf947bb.js';
import '../../../_chunks/dep-d574012b.js';
import '../../../_chunks/dep-291b3520.js';
import '../../../_chunks/dep-43f2594d.js';
import '../../../_chunks/dep-dd510963.js';
import '../../../_chunks/dep-ab257f91.js';
import '../../../_chunks/dep-d8abc91d.js';
import '../../../_chunks/dep-e852e0d4.js';
import '../../../_chunks/dep-c67c8434.js';
import '../../../_chunks/dep-61447210.js';
import '../../../_chunks/dep-50ef74ba.js';
import '../../../_chunks/dep-42e9f2d1.js';
import '../../../_chunks/dep-ff001f02.js';
import '../../../_chunks/dep-1ff7016b.js';
import '../../../_chunks/dep-7dbcd8fa.js';
import '../../../_chunks/dep-213a15b4.js';
import '../../../_chunks/dep-f35f0cd8.js';
import '../../../_chunks/dep-ee09d461.js';
import '../../../_chunks/dep-48a2ba1f.js';
import '../../../_chunks/dep-373953f1.js';
import '../../../_chunks/dep-76e94c47.js';
import '../../../_chunks/dep-39df2a76.js';
import '../../../_chunks/dep-06ad19cb.js';
import '../../../_chunks/dep-02fa0b93.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function getExposedProps(node) {
  var props = pick_1(node, ["value", "label", "data", "actived", "expanded", "checked", "indeterminate", "loading"]);
  return props;
}
function createNodeModel(node) {
  var props = getExposedProps(node);
  var model = _objectSpread(_objectSpread({}, props), {}, {
    getLevel: function getLevel() {
      return node.getLevel();
    },
    getIndex: function getIndex() {
      return node.getIndex();
    },
    isFirst: function isFirst() {
      return node.isFirst();
    },
    isLast: function isLast() {
      return node.isLast();
    },
    isLeaf: function isLeaf() {
      return node.isLeaf();
    },
    insertBefore: function insertBefore(newData) {
      return node.insertBefore(newData);
    },
    insertAfter: function insertAfter(newData) {
      return node.insertAfter(newData);
    },
    appendData: function appendData(data) {
      return node.append(data);
    },
    getPath: function getPath() {
      var nodes = node.getPath();
      return nodes.map(function (item) {
        return item.getModel();
      });
    },
    getParent: function getParent() {
      var _node$parent;
      return (_node$parent = node.parent) === null || _node$parent === void 0 ? void 0 : _node$parent.getModel();
    },
    getParents: function getParents() {
      var nodes = node.getParents();
      return nodes.map(function (item) {
        return item.getModel();
      });
    },
    getRoot: function getRoot() {
      var root = node.getRoot();
      return root === null || root === void 0 ? void 0 : root.getModel();
    },
    getSiblings: function getSiblings() {
      var nodes = node.getSiblings();
      return nodes.map(function (item) {
        return item.getModel();
      });
    },
    getChildren: function getChildren(deep) {
      var childrenModel = false;
      var children = node.children;
      if (Array.isArray(children)) {
        if (children.length > 0) {
          if (deep) {
            var nodes = node.walk();
            nodes.shift();
            childrenModel = nodes.map(function (item) {
              return item.getModel();
            });
          } else {
            childrenModel = children.map(function (item) {
              return item.getModel();
            });
          }
        } else {
          childrenModel = false;
        }
      } else if (isBoolean_1(children)) {
        childrenModel = children;
      }
      return childrenModel;
    },
    remove: function remove(value) {
      if (!value) {
        node.remove();
        return;
      }
      var tree = node.tree;
      var targetNode = tree.getNode(value);
      if (!targetNode) {
        log.warnOnce("Tree", "`".concat(value, "` is not exist"));
        return;
      }
      var parents = targetNode.getParents();
      var parentValues = parents.map(function (pnode) {
        return pnode.value;
      });
      if (parentValues.indexOf(node.value) < 0) {
        log.warnOnce("Tree", "`".concat(value, "` is not a childNode of current node"));
        return;
      }
      targetNode.remove();
    },
    setData: function setData(data) {
      var _data = omit_1(data, ["children", "value", "label"]);
      var keys = node.tree.config.keys;
      var dataValue = data[(keys === null || keys === void 0 ? void 0 : keys.value) || "value"];
      var dataLabel = data[(keys === null || keys === void 0 ? void 0 : keys.label) || "label"];
      if (!isUndefined_1(dataValue)) _data.value = dataValue;
      if (!isUndefined_1(dataLabel)) _data.label = dataLabel;
      Object.assign(node.data, _data);
      Object.assign(node, _data);
    }
  });
  return model;
}
function updateNodeModel(model, node) {
  var props = getExposedProps(node);
  Object.assign(model, props);
}

export { createNodeModel, updateNodeModel };
