// tdesign-react/lib/_common/js/tree-v1/tree-node-model.js
/**
 * tdesign v1.7.1
 * (c) 2024 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../../../_chunks/dep-9d06ffd4.js';
import { i as isUndefined_1 } from '../../../_chunks/dep-1bcd89bc.js';
import { i as isBoolean_1 } from '../../../_chunks/dep-5f785761.js';
import { p as pick_1 } from '../../../_chunks/dep-f94e0858.js';
import { o as omit_1 } from '../../../_chunks/dep-1bd754c9.js';
import log from '../log/log.js';
import '../../../_chunks/dep-f3f3df31.js';
import '../../../_chunks/dep-9d4210ce.js';
import '../../../_chunks/dep-bee1cff5.js';
import '../../../_chunks/dep-d5932774.js';
import '../../../_chunks/dep-e6da4f4e.js';
import '../../../_chunks/dep-933fa6aa.js';
import '../../../_chunks/dep-f1d40d68.js';
import '../../../_chunks/dep-5f0ec823.js';
import '../../../_chunks/dep-44c20420.js';
import '../../../_chunks/dep-f3c9f0b1.js';
import '../../../_chunks/dep-3317a9f1.js';
import '../../../_chunks/dep-33f13f9c.js';
import '../../../_chunks/dep-23dea7e8.js';
import '../../../_chunks/dep-cf0c62b9.js';
import '../../../_chunks/dep-560d448e.js';
import '../../../_chunks/dep-d5964765.js';
import '../../../_chunks/dep-f8d0dc23.js';
import '../../../_chunks/dep-a1b2bf89.js';
import '../../../_chunks/dep-7b4b8732.js';
import '../../../_chunks/dep-03c231fb.js';
import '../../../_chunks/dep-f6eefc4b.js';
import '../../../_chunks/dep-63a5ea33.js';
import '../../../_chunks/dep-88ade432.js';
import '../../../_chunks/dep-d841f939.js';
import '../../../_chunks/dep-8e2ab2c9.js';
import '../../../_chunks/dep-0064325f.js';
import '../../../_chunks/dep-ae0cbd35.js';
import '../../../_chunks/dep-9823fa03.js';
import '../../../_chunks/dep-a5a9bbb3.js';
import '../../../_chunks/dep-71b520ec.js';
import '../../../_chunks/dep-02774c4b.js';
import '../../../_chunks/dep-b4f76a42.js';
import '../../../_chunks/dep-bcd6d621.js';
import '../../../_chunks/dep-c464989f.js';
import '../../../_chunks/dep-cf209156.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function getExposedProps(node) {
  var props = pick_1(node, ["value", "label", "data", "actived", "expanded", "checked", "indeterminate", "loading"]);
  return props;
}
function createNodeModel(node) {
  var props = getExposedProps(node);
  var model = _objectSpread(_objectSpread({}, props), {}, {
    getLevel: function getLevel() {
      return node.getLevel();
    },
    getIndex: function getIndex() {
      return node.getIndex();
    },
    isFirst: function isFirst() {
      return node.isFirst();
    },
    isLast: function isLast() {
      return node.isLast();
    },
    isLeaf: function isLeaf() {
      return node.isLeaf();
    },
    insertBefore: function insertBefore(newData) {
      return node.insertBefore(newData);
    },
    insertAfter: function insertAfter(newData) {
      return node.insertAfter(newData);
    },
    appendData: function appendData(data) {
      return node.append(data);
    },
    getPath: function getPath() {
      var nodes = node.getPath();
      return nodes.map(function (item) {
        return item.getModel();
      });
    },
    getParent: function getParent() {
      var _node$parent;
      return (_node$parent = node.parent) === null || _node$parent === void 0 ? void 0 : _node$parent.getModel();
    },
    getParents: function getParents() {
      var nodes = node.getParents();
      return nodes.map(function (item) {
        return item.getModel();
      });
    },
    getRoot: function getRoot() {
      var root = node.getRoot();
      return root === null || root === void 0 ? void 0 : root.getModel();
    },
    getSiblings: function getSiblings() {
      var nodes = node.getSiblings();
      return nodes.map(function (item) {
        return item.getModel();
      });
    },
    getChildren: function getChildren(deep) {
      var childrenModel = false;
      var children = node.children;
      if (Array.isArray(children)) {
        if (children.length > 0) {
          if (deep) {
            var nodes = node.walk();
            nodes.shift();
            childrenModel = nodes.map(function (item) {
              return item.getModel();
            });
          } else {
            childrenModel = children.map(function (item) {
              return item.getModel();
            });
          }
        } else {
          childrenModel = false;
        }
      } else if (isBoolean_1(children)) {
        childrenModel = children;
      }
      return childrenModel;
    },
    remove: function remove(value) {
      if (!value) {
        node.remove();
        return;
      }
      var tree = node.tree;
      var targetNode = tree.getNode(value);
      if (!targetNode) {
        log.warnOnce("Tree", "`".concat(value, "` is not exist"));
        return;
      }
      var parents = targetNode.getParents();
      var parentValues = parents.map(function (pnode) {
        return pnode.value;
      });
      if (parentValues.indexOf(node.value) < 0) {
        log.warnOnce("Tree", "`".concat(value, "` is not a childNode of current node"));
        return;
      }
      targetNode.remove();
    },
    setData: function setData(data) {
      var _data = omit_1(data, ["children", "value", "label"]);
      var keys = node.tree.config.keys;
      var dataValue = data[(keys === null || keys === void 0 ? void 0 : keys.value) || "value"];
      var dataLabel = data[(keys === null || keys === void 0 ? void 0 : keys.label) || "label"];
      if (!isUndefined_1(dataValue)) _data.value = dataValue;
      if (!isUndefined_1(dataLabel)) _data.label = dataLabel;
      Object.assign(node.data, _data);
      Object.assign(node, _data);
    }
  });
  return model;
}
function updateNodeModel(model, node) {
  var props = getExposedProps(node);
  Object.assign(model, props);
}

export { createNodeModel, updateNodeModel };
