// tdesign-react/cjs/_common/js/tree-v1/tree-node-model.js
/**
 * tdesign v1.7.1
 * (c) 2024 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var defineProperty = require('../../../_chunks/dep-ff9a5a06.js');
var isUndefined = require('../../../_chunks/dep-0f313135.js');
var isBoolean = require('../../../_chunks/dep-f4b44bcb.js');
var pick = require('../../../_chunks/dep-e0e28478.js');
var omit = require('../../../_chunks/dep-db8a90e2.js');
var _common_js_log_log = require('../log/log.js');
require('../../../_chunks/dep-a4fb1294.js');
require('../../../_chunks/dep-68d6ca4f.js');
require('../../../_chunks/dep-59101a1c.js');
require('../../../_chunks/dep-315a4a2b.js');
require('../../../_chunks/dep-55437cd3.js');
require('../../../_chunks/dep-e7a0e8b4.js');
require('../../../_chunks/dep-f70c56e2.js');
require('../../../_chunks/dep-4a411278.js');
require('../../../_chunks/dep-9171afa5.js');
require('../../../_chunks/dep-7d38d5e1.js');
require('../../../_chunks/dep-5ed3b3ae.js');
require('../../../_chunks/dep-8bb91473.js');
require('../../../_chunks/dep-04f46890.js');
require('../../../_chunks/dep-2a42b204.js');
require('../../../_chunks/dep-d93c04ca.js');
require('../../../_chunks/dep-2a8c4619.js');
require('../../../_chunks/dep-6bae679a.js');
require('../../../_chunks/dep-1c485dd2.js');
require('../../../_chunks/dep-3076c3e7.js');
require('../../../_chunks/dep-64e9d9c5.js');
require('../../../_chunks/dep-ef5f8027.js');
require('../../../_chunks/dep-66fdf975.js');
require('../../../_chunks/dep-33b4bb35.js');
require('../../../_chunks/dep-2a862cb7.js');
require('../../../_chunks/dep-d9f1fc69.js');
require('../../../_chunks/dep-2caefa8e.js');
require('../../../_chunks/dep-14dd074a.js');
require('../../../_chunks/dep-e38461b1.js');
require('../../../_chunks/dep-519d75c6.js');
require('../../../_chunks/dep-8685fcba.js');
require('../../../_chunks/dep-ef8352c0.js');
require('../../../_chunks/dep-1dbff57a.js');
require('../../../_chunks/dep-f74209a9.js');
require('../../../_chunks/dep-174c24e3.js');
require('../../../_chunks/dep-bd15ffce.js');

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { defineProperty._defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function getExposedProps(node) {
  var props = pick.pick_1(node, ["value", "label", "data", "actived", "expanded", "checked", "indeterminate", "loading"]);
  return props;
}
function createNodeModel(node) {
  var props = getExposedProps(node);
  var model = _objectSpread(_objectSpread({}, props), {}, {
    getLevel: function getLevel() {
      return node.getLevel();
    },
    getIndex: function getIndex() {
      return node.getIndex();
    },
    isFirst: function isFirst() {
      return node.isFirst();
    },
    isLast: function isLast() {
      return node.isLast();
    },
    isLeaf: function isLeaf() {
      return node.isLeaf();
    },
    insertBefore: function insertBefore(newData) {
      return node.insertBefore(newData);
    },
    insertAfter: function insertAfter(newData) {
      return node.insertAfter(newData);
    },
    appendData: function appendData(data) {
      return node.append(data);
    },
    getPath: function getPath() {
      var nodes = node.getPath();
      return nodes.map(function (item) {
        return item.getModel();
      });
    },
    getParent: function getParent() {
      var _node$parent;
      return (_node$parent = node.parent) === null || _node$parent === void 0 ? void 0 : _node$parent.getModel();
    },
    getParents: function getParents() {
      var nodes = node.getParents();
      return nodes.map(function (item) {
        return item.getModel();
      });
    },
    getRoot: function getRoot() {
      var root = node.getRoot();
      return root === null || root === void 0 ? void 0 : root.getModel();
    },
    getSiblings: function getSiblings() {
      var nodes = node.getSiblings();
      return nodes.map(function (item) {
        return item.getModel();
      });
    },
    getChildren: function getChildren(deep) {
      var childrenModel = false;
      var children = node.children;
      if (Array.isArray(children)) {
        if (children.length > 0) {
          if (deep) {
            var nodes = node.walk();
            nodes.shift();
            childrenModel = nodes.map(function (item) {
              return item.getModel();
            });
          } else {
            childrenModel = children.map(function (item) {
              return item.getModel();
            });
          }
        } else {
          childrenModel = false;
        }
      } else if (isBoolean.isBoolean_1(children)) {
        childrenModel = children;
      }
      return childrenModel;
    },
    remove: function remove(value) {
      if (!value) {
        node.remove();
        return;
      }
      var tree = node.tree;
      var targetNode = tree.getNode(value);
      if (!targetNode) {
        _common_js_log_log["default"].warnOnce("Tree", "`".concat(value, "` is not exist"));
        return;
      }
      var parents = targetNode.getParents();
      var parentValues = parents.map(function (pnode) {
        return pnode.value;
      });
      if (parentValues.indexOf(node.value) < 0) {
        _common_js_log_log["default"].warnOnce("Tree", "`".concat(value, "` is not a childNode of current node"));
        return;
      }
      targetNode.remove();
    },
    setData: function setData(data) {
      var _data = omit.omit_1(data, ["children", "value", "label"]);
      var keys = node.tree.config.keys;
      var dataValue = data[(keys === null || keys === void 0 ? void 0 : keys.value) || "value"];
      var dataLabel = data[(keys === null || keys === void 0 ? void 0 : keys.label) || "label"];
      if (!isUndefined.isUndefined_1(dataValue)) _data.value = dataValue;
      if (!isUndefined.isUndefined_1(dataLabel)) _data.label = dataLabel;
      Object.assign(node.data, _data);
      Object.assign(node, _data);
    }
  });
  return model;
}
function updateNodeModel(model, node) {
  var props = getExposedProps(node);
  Object.assign(model, props);
}

exports.createNodeModel = createNodeModel;
exports.updateNodeModel = updateNodeModel;
