// tdesign-react/es/table/hooks/useTreeDataExpand.js
/**
 * tdesign v1.7.1
 * (c) 2024 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../../_chunks/dep-9d06ffd4.js';
import { _ as _toConsumableArray } from '../../_chunks/dep-1336b163.js';
import { _ as _slicedToArray } from '../../_chunks/dep-be37289b.js';
import { useState, useEffect } from 'react';
import usePrevious from '../../hooks/usePrevious.js';
import useControlled from '../../hooks/useControlled.js';
import { getUniqueRowValue, diffExpandedTreeNode } from '../../_common/js/table/tree-store.js';
import '../../_chunks/dep-e0a22ada.js';
import '../../_chunks/dep-0163b4c6.js';
import '../../_chunks/dep-08fc9b22.js';
import '../../_chunks/dep-199643a3.js';
import '../../_chunks/dep-06ad19cb.js';
import '../../_chunks/dep-d574012b.js';
import '../../_chunks/dep-f3bac8f4.js';
import '../../_chunks/dep-9d4210ce.js';
import '../../_chunks/dep-17c2aa2b.js';
import '../../_chunks/dep-c3a67e8f.js';
import '../../_chunks/dep-dc4499f2.js';
import '../../_util/noop.js';
import '../../_chunks/dep-174657df.js';
import '../../_chunks/dep-f2dd6554.js';
import '../../_chunks/dep-fdd4f070.js';
import '../../_chunks/dep-49b9df8b.js';
import '../../_chunks/dep-cc3b0635.js';
import '../../_chunks/dep-39d65f32.js';
import '../../_chunks/dep-70de8f55.js';
import '../../_chunks/dep-c2c3c04d.js';
import '../../_chunks/dep-48a2bd8e.js';
import '../../_chunks/dep-8cf947bb.js';
import '../../_chunks/dep-9c770917.js';
import '../../_chunks/dep-291b3520.js';
import '../../_chunks/dep-43f2594d.js';
import '../../_chunks/dep-dd510963.js';
import '../../_chunks/dep-ab257f91.js';
import '../../_common/js/table/utils.js';
import '../../_chunks/dep-05f57d79.js';
import '../../_common/js/log/index.js';
import '../../_common/js/log/log.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function useTreeDataExpand(props, params) {
  var store = params.store,
    dataSource = params.dataSource,
    rowDataKeys = params.rowDataKeys,
    setDataSource = params.setDataSource;
  var data = props.data,
    tree = props.tree;
  var _useState = useState(false),
    _useState2 = _slicedToArray(_useState, 2),
    isDefaultExpandAllExecute = _useState2[0],
    setIsDefaultExpandAllExecute = _useState2[1];
  var _useControlled = useControlled(props, "expandedTreeNodes", props.onExpandedTreeNodesChange, {
      defaultExpandedTreeNodes: props.defaultSelectedRowKeys || []
    }),
    _useControlled2 = _slicedToArray(_useControlled, 2),
    tExpandedTreeNode = _useControlled2[0],
    setTExpandedTreeNode = _useControlled2[1];
  var oldExpandedTreeNode = usePrevious(tExpandedTreeNode);
  var _useState3 = useState({
      type: "props-change"
    }),
    _useState4 = _slicedToArray(_useState3, 2),
    changedExpandTreeNode = _useState4[0],
    setChangedExpandTreeNode = _useState4[1];
  function expandAll() {
    var type = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : "expand-all";
    var list = arguments.length > 1 ? arguments[1] : undefined;
    var newData = list || data;
    setDataSource(store.expandAll(newData, rowDataKeys));
    var expandedNode = dataSource.map(function (t) {
      return getUniqueRowValue(t, rowDataKeys.rowKey);
    });
    setTExpandedTreeNode(expandedNode, {
      row: void 0,
      rowState: void 0,
      rowIndex: void 0,
      type: "expand",
      trigger: type
    });
    setChangedExpandTreeNode({
      type: "expand-all"
    });
  }
  function foldAll() {
    setDataSource(_toConsumableArray(store.foldAll(dataSource, rowDataKeys)));
    setTExpandedTreeNode([], {
      row: void 0,
      rowState: void 0,
      rowIndex: void 0,
      type: "fold",
      trigger: "fold-all"
    });
  }
  function onExpandFoldIconClick(p, trigger) {
    var _props$onTreeExpandCh;
    var row = p.row,
      rowIndex = p.rowIndex;
    setChangedExpandTreeNode(_objectSpread({
      type: "user-reaction-change"
    }, p));
    var rowValue = getUniqueRowValue(row, rowDataKeys.rowKey);
    var rowState = store.treeDataMap.get(rowValue);
    var expandedNodes = _toConsumableArray(tExpandedTreeNode);
    if (rowState.expanded) {
      var expandedChildrenKeys = store.getExpandedChildrenKeys([row], rowDataKeys);
      for (var i = 0, len = expandedNodes.length; i < len; i++) {
        var nodeValue = expandedNodes[i];
        if (expandedChildrenKeys.includes(nodeValue)) {
          expandedNodes[i] = void 0;
        }
      }
      expandedNodes = expandedNodes.filter(Boolean);
    } else {
      expandedNodes.push(rowValue);
    }
    var params2 = {
      row: row,
      rowIndex: rowIndex,
      rowState: rowState,
      trigger: trigger
    };
    setTExpandedTreeNode(expandedNodes, _objectSpread(_objectSpread({}, params2), {}, {
      type: rowState.expanded ? "fold" : "expand"
    }));
    (_props$onTreeExpandCh = props.onTreeExpandChange) === null || _props$onTreeExpandCh === void 0 || _props$onTreeExpandCh.call(props, params2);
  }
  function updateExpandState(data2, tExpandedTreeNode2) {
    var oldExpandedTreeNode2 = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : [];
    var _diffExpandedTreeNode = diffExpandedTreeNode(tExpandedTreeNode2, oldExpandedTreeNode2),
      addedList = _diffExpandedTreeNode.addedList,
      removedList = _diffExpandedTreeNode.removedList;
    store.expandTreeNode(addedList, data2, rowDataKeys);
    store.foldTreeNode(removedList, data2, rowDataKeys);
    return data2;
  }
  useEffect(function () {
    if (!store.treeDataMap.size) return;
    if (changedExpandTreeNode.type === "user-reaction-change") {
      var _ref = changedExpandTreeNode || {},
        row = _ref.row,
        rowIndex = _ref.rowIndex;
      var newData = store.toggleExpandData({
        row: row,
        rowIndex: rowIndex
      }, dataSource, rowDataKeys);
      setDataSource(_toConsumableArray(newData));
    } else if (changedExpandTreeNode.type === "props-change") {
      updateExpandState(_toConsumableArray(dataSource), tExpandedTreeNode, oldExpandedTreeNode);
    }
    if (changedExpandTreeNode.type !== "props-change") {
      setChangedExpandTreeNode({
        type: "props-change"
      });
    }
  }, [tExpandedTreeNode]);
  var updateExpandOnDataChange = function updateExpandOnDataChange(data2) {
    if (tree !== null && tree !== void 0 && tree.defaultExpandAll && !isDefaultExpandAllExecute) {
      expandAll("default-expand-all", _toConsumableArray(data2));
      setIsDefaultExpandAllExecute(true);
    } else if (tExpandedTreeNode !== null && tExpandedTreeNode !== void 0 && tExpandedTreeNode.length) {
      var newData = updateExpandState(_toConsumableArray(data2), tExpandedTreeNode, []);
      setDataSource(_toConsumableArray(newData));
    }
  };
  return {
    tExpandedTreeNode: tExpandedTreeNode,
    isDefaultExpandAllExecute: isDefaultExpandAllExecute,
    expandAll: expandAll,
    foldAll: foldAll,
    onExpandFoldIconClick: onExpandFoldIconClick,
    updateExpandOnDataChange: updateExpandOnDataChange
  };
}

export { useTreeDataExpand as default, useTreeDataExpand };
