// tdesign-react/es/sticky-tool/StickyItem.js
/**
 * tdesign v1.7.1
 * (c) 2024 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-9d06ffd4.js';
import React, { useMemo, useCallback } from 'react';
import classNames from 'classnames';
import useConfig from '../hooks/useConfig.js';
import { Popup } from '../popup/index.js';
import '../config-provider/ConfigContext.js';
import '../_chunks/dep-f870baf9.js';
import '../_chunks/dep-08fc9b22.js';
import '../_chunks/dep-1ff7016b.js';
import '../_chunks/dep-cc3b0635.js';
import '../_chunks/dep-39d65f32.js';
import '../_chunks/dep-70de8f55.js';
import '../_chunks/dep-f3bac8f4.js';
import '../_chunks/dep-9d4210ce.js';
import '../_chunks/dep-c2c3c04d.js';
import '../_chunks/dep-48a2bd8e.js';
import '../_chunks/dep-8cf947bb.js';
import '../_chunks/dep-7dbcd8fa.js';
import '../_chunks/dep-e852e0d4.js';
import '../_chunks/dep-dc4499f2.js';
import '../_chunks/dep-213a15b4.js';
import '../_chunks/dep-43f2594d.js';
import '../_chunks/dep-dd510963.js';
import '../_chunks/dep-d8abc91d.js';
import '../_chunks/dep-17c2aa2b.js';
import '../_chunks/dep-ab257f91.js';
import '../_chunks/dep-d4455e7d.js';
import '../_chunks/dep-42e9f2d1.js';
import '../_chunks/dep-02fa0b93.js';
import '../_chunks/dep-24058d76.js';
import '../locale/zh_CN.js';
import '../_common/js/global-config/locale/zh_CN.js';
import '../_chunks/dep-0f18acb0.js';
import 'dayjs';
import '../_chunks/dep-8b683475.js';
import '../_common/js/global-config/default-config.js';
import '../popup/Popup.js';
import '../_chunks/dep-be37289b.js';
import '../_chunks/dep-e0a22ada.js';
import 'react-transition-group';
import '../_chunks/dep-5b468636.js';
import '../_chunks/dep-b6cd9ebe.js';
import '../_chunks/dep-c3a67e8f.js';
import 'react-popper';
import '../hooks/useControlled.js';
import '../_chunks/dep-0163b4c6.js';
import '../_chunks/dep-199643a3.js';
import '../_chunks/dep-06ad19cb.js';
import '../_chunks/dep-d574012b.js';
import '../_util/noop.js';
import '../_util/useAnimation.js';
import '../common/Portal.js';
import 'react-dom';
import '../_util/dom.js';
import 'raf';
import '../_chunks/dep-3f719063.js';
import '../_util/easing.js';
import '../popup/hooks/useTrigger.js';
import 'react-is';
import '../popup/utils/ref.js';
import '../_util/composeRefs.js';
import '../popup/utils/transition.js';
import '../_util/useMutationObserver.js';
import '../_util/useWindowSize.js';
import '../popup/defaultProps.js';
import '../hooks/useDefaultProps.js';
import '../popup/PopupPlugin.js';
import '../_chunks/dep-57f1d6d7.js';
import '@popperjs/core';
import '../_util/react-render.js';
import '../popup/style/css.js';
import '../popup/type.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var StickyItem = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var icon = props.icon,
    label = props.label,
    popup = props.popup,
    popupProps = props.popupProps,
    trigger = props.trigger,
    type = props.type,
    shape = props.shape,
    placement = props.placement,
    basePopupProps = props.basePopupProps,
    baseWidth = props.baseWidth,
    onClick = props.onClick,
    onHover = props.onHover,
    style = props.style,
    className = props.className;
  var _useConfig = useConfig(),
    classPrefix = _useConfig.classPrefix;
  var popupPlacement = useMemo(function () {
    return placement.includes("right") ? "left" : "right";
  }, [placement]);
  var styles = useMemo(function () {
    var styles2 = _objectSpread({}, style);
    if (baseWidth) {
      var selfWidth = type === "normal" ? "56px" : "40px";
      styles2.margin = "calc((".concat(baseWidth, " - ").concat(selfWidth, ")/2)");
    }
    return styles2;
  }, [baseWidth, style, type]);
  var handleClickItem = useCallback(function (e) {
    var item = {
      icon: icon,
      label: label,
      popup: popup,
      popupProps: popupProps,
      trigger: trigger
    };
    onClick({
      e: e,
      item: item
    });
  }, [icon, label, popup, popupProps, trigger, onClick]);
  var handleHoverItem = useCallback(function (e) {
    var item = {
      icon: icon,
      label: label,
      popup: popup,
      popupProps: popupProps,
      trigger: trigger
    };
    onHover({
      e: e,
      item: item
    });
  }, [icon, label, popup, popupProps, trigger, onHover]);
  var finalPopupProps = _objectSpread(_objectSpread({
    hideEmptyPopup: true
  }, basePopupProps), popupProps);
  return /* @__PURE__ */React.createElement(Popup, _objectSpread({
    overlayInnerClassName: classNames("".concat(classPrefix, "-sticky-tool-popup-content")),
    trigger: props.trigger,
    hideEmptyPopup: true,
    placement: popupPlacement,
    content: popup
  }, finalPopupProps), /* @__PURE__ */React.createElement("div", {
    ref: ref,
    className: classNames("".concat(classPrefix, "-sticky-item"), "".concat(classPrefix, "-sticky-item--").concat(shape), "".concat(classPrefix, "-sticky-item--").concat(type), className),
    style: styles,
    onClick: handleClickItem,
    onMouseEnter: handleHoverItem
  }, icon, props.type === "normal" ? /* @__PURE__ */React.createElement("div", {
    className: classNames("".concat(classPrefix, "-sticky-item__label"))
  }, label) : null));
});
StickyItem.displayName = "StickyItem";

export { StickyItem as default };
