// tdesign-react/esm/_chunks/dep-88cead75.js
/**
 * tdesign v1.7.1
 * (c) 2024 tdesign
 * @license MIT
 */

import React, { useMemo } from 'react';
import './dep-08fc9b22.js';
import { t as toString_1 } from './dep-d574012b.js';
import useConfig from '../hooks/useConfig.js';

var toString = toString_1;

/**
 * Used to match `RegExp`
 * [syntax characters](http://ecma-international.org/ecma-262/7.0/#sec-patterns).
 */
var reRegExpChar = /[\\^$.*+?()[\]{}|]/g,
  reHasRegExpChar = RegExp(reRegExpChar.source);

/**
 * Escapes the `RegExp` special characters "^", "$", "\", ".", "*", "+",
 * "?", "(", ")", "[", "]", "{", "}", and "|" in `string`.
 *
 * @static
 * @memberOf _
 * @since 3.0.0
 * @category String
 * @param {string} [string=''] The string to escape.
 * @returns {string} Returns the escaped string.
 * @example
 *
 * _.escapeRegExp('[lodash](https://lodash.com/)');
 * // => '\[lodash\]\(https://lodash\.com/\)'
 */
function escapeRegExp(string) {
  string = toString(string);
  return string && reHasRegExpChar.test(string) ? string.replace(reRegExpChar, '\\$&') : string;
}
var escapeRegExp_1 = escapeRegExp;

var HighlightOption = function HighlightOption(props) {
  var _useConfig = useConfig(),
    classPrefix = _useConfig.classPrefix;
  var content = props.content,
    keyword = props.keyword;
  var words = useMemo(function () {
    var _content$match;
    if (!content) return {
      list: []
    };
    if (typeof content !== "string" || !keyword) return {
      list: [content]
    };
    var regExp = new RegExp(escapeRegExp_1(keyword), "i");
    var splitKeyword = (_content$match = content.match(regExp)) === null || _content$match === void 0 ? void 0 : _content$match[0];
    return {
      list: content.split(splitKeyword),
      keyword: splitKeyword
    };
  }, [content, keyword]);
  return /* @__PURE__ */React.createElement("div", {
    className: "".concat(classPrefix, "-select-option__highlight-item")
  }, words.list.map(function (item, index) {
    if (!index) return item;
    return [/* @__PURE__ */React.createElement("b", {
      className: "".concat(classPrefix, "-is-highlight"),
      key: item + words.keyword
    }, words.keyword), item];
  }));
};
HighlightOption.displayName = "HighlightOption";

export { HighlightOption as H, escapeRegExp_1 as e };
