// tdesign-react/esm/_util/useCommonClassName.js
/**
 * tdesign v1.7.1
 * (c) 2024 tdesign
 * @license MIT
 */

import { useMemo } from 'react';
import useConfig from '../hooks/useConfig.js';
import '../config-provider/ConfigContext.js';
import '../_chunks/dep-9d06ffd4.js';
import '../_chunks/dep-f870baf9.js';
import '../_chunks/dep-08fc9b22.js';
import '../_chunks/dep-1ff7016b.js';
import '../_chunks/dep-cc3b0635.js';
import '../_chunks/dep-39d65f32.js';
import '../_chunks/dep-70de8f55.js';
import '../_chunks/dep-f3bac8f4.js';
import '../_chunks/dep-9d4210ce.js';
import '../_chunks/dep-c2c3c04d.js';
import '../_chunks/dep-48a2bd8e.js';
import '../_chunks/dep-8cf947bb.js';
import '../_chunks/dep-7dbcd8fa.js';
import '../_chunks/dep-e852e0d4.js';
import '../_chunks/dep-dc4499f2.js';
import '../_chunks/dep-213a15b4.js';
import '../_chunks/dep-43f2594d.js';
import '../_chunks/dep-dd510963.js';
import '../_chunks/dep-d8abc91d.js';
import '../_chunks/dep-17c2aa2b.js';
import '../_chunks/dep-ab257f91.js';
import '../_chunks/dep-d4455e7d.js';
import '../_chunks/dep-42e9f2d1.js';
import '../_chunks/dep-02fa0b93.js';
import '../_chunks/dep-24058d76.js';
import '../locale/zh_CN.js';
import '../_common/js/global-config/locale/zh_CN.js';
import '../_chunks/dep-0f18acb0.js';
import 'dayjs';
import '../_chunks/dep-8b683475.js';
import '../_common/js/global-config/default-config.js';

function useCommonClassName() {
  var _useConfig = useConfig(),
    classPrefix = _useConfig.classPrefix;
  return useMemo(function () {
    return {
      SIZE: {
        "default": "",
        xs: "".concat(classPrefix, "-size-xs"),
        small: "".concat(classPrefix, "-size-s"),
        medium: "".concat(classPrefix, "-size-m"),
        large: "".concat(classPrefix, "-size-l"),
        xl: "".concat(classPrefix, "-size-xl"),
        block: "".concat(classPrefix, "-size-full-width")
      },
      STATUS: {
        loading: "".concat(classPrefix, "-is-loading"),
        disabled: "".concat(classPrefix, "-is-disabled"),
        focused: "".concat(classPrefix, "-is-focused"),
        success: "".concat(classPrefix, "-is-success"),
        error: "".concat(classPrefix, "-is-error"),
        warning: "".concat(classPrefix, "-is-warning"),
        selected: "".concat(classPrefix, "-is-selected"),
        active: "".concat(classPrefix, "-is-active"),
        checked: "".concat(classPrefix, "-is-checked"),
        current: "".concat(classPrefix, "-is-current"),
        hidden: "".concat(classPrefix, "-is-hidden"),
        visible: "".concat(classPrefix, "-is-visible"),
        expanded: "".concat(classPrefix, "-is-expanded"),
        indeterminate: "".concat(classPrefix, "-is-indeterminate")
      }
    };
  }, [classPrefix]);
}

export { useCommonClassName as default };
