// tdesign-react/lib/select/hooks/useOptions.js
/**
 * tdesign v1.7.1
 * (c) 2024 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../../_chunks/dep-9d06ffd4.js';
import { _ as _slicedToArray } from '../../_chunks/dep-be37289b.js';
import React, { useState, useEffect } from 'react';
import { g as get_1 } from '../../_chunks/dep-92001d50.js';
import { getValueToOption } from '../util/helper.js';
import Option from '../base/Option.js';
import '../../_chunks/dep-e0a22ada.js';
import '../../_chunks/dep-e6da4f4e.js';
import '../../_chunks/dep-933fa6aa.js';
import '../../_chunks/dep-9d4210ce.js';
import '../../_chunks/dep-f1d40d68.js';
import '../../_chunks/dep-f3f3df31.js';
import '../../_chunks/dep-bee1cff5.js';
import '../../_chunks/dep-d5932774.js';
import '../../_chunks/dep-5f0ec823.js';
import '../../_chunks/dep-44c20420.js';
import '../../_chunks/dep-f3c9f0b1.js';
import '../../_chunks/dep-3317a9f1.js';
import '../../_chunks/dep-33f13f9c.js';
import '../../_chunks/dep-23dea7e8.js';
import '../../_chunks/dep-1336b163.js';
import '../../_chunks/dep-cf209156.js';
import '../../_chunks/dep-8e2ab2c9.js';
import '../../_chunks/dep-0064325f.js';
import '../../_chunks/dep-7b4b8732.js';
import '../../_chunks/dep-ae0cbd35.js';
import '../../_chunks/dep-560d448e.js';
import '../../_chunks/dep-d5964765.js';
import '../../_chunks/dep-a1b2bf89.js';
import '../../_chunks/dep-f8d0dc23.js';
import '../base/OptionGroup.js';
import 'classnames';
import '../../hooks/useConfig.js';
import '../../config-provider/ConfigContext.js';
import '../../_chunks/dep-24ea85d3.js';
import '../../_chunks/dep-434ba20b.js';
import '../../_chunks/dep-88ade432.js';
import '../../_chunks/dep-dc6bd1bf.js';
import '../../_common/js/global-config/locale/zh_CN.js';
import '../../_chunks/dep-7ba5fe7d.js';
import 'dayjs';
import '../../_common/js/global-config/default-config.js';
import '../defaultProps.js';
import '../../hooks/useDefaultProps.js';
import '../../_chunks/dep-69145a1d.js';
import '../../_chunks/dep-c6a70169.js';
import '../../hooks/useDomRefCallback.js';
import '../../_util/useRipple.js';
import '../../_util/useAnimation.js';
import '../../_util/setStyle.js';
import '../../_util/dom.js';
import 'raf';
import '../../_util/easing.js';

function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { _defineProperty(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function UseOptions(keys, options, children, valueType, value, reserveKeyword) {
  var _useState = useState({}),
    _useState2 = _slicedToArray(_useState, 2),
    valueToOption = _useState2[0],
    setValueToOption = _useState2[1];
  var _useState3 = useState([]),
    _useState4 = _slicedToArray(_useState3, 2),
    currentOptions = _useState4[0],
    setCurrentOptions = _useState4[1];
  var _useState5 = useState([]),
    _useState6 = _slicedToArray(_useState5, 2),
    tmpPropOptions = _useState6[0],
    setTmpPropOptions = _useState6[1];
  var _useState7 = useState([]),
    _useState8 = _slicedToArray(_useState7, 2),
    selectedOptions = _useState8[0],
    setSelectedOptions = _useState8[1];
  useEffect(function () {
    var transformedOptions = options;
    var arrayChildren = React.Children.toArray(children);
    var optionChildren = arrayChildren.filter(function (v) {
      return v.type === Option;
    });
    var isChildrenFilterable = arrayChildren.length > 0 && optionChildren.length === arrayChildren.length;
    if (reserveKeyword && currentOptions.length && isChildrenFilterable) return;
    if (isChildrenFilterable) {
      transformedOptions = arrayChildren === null || arrayChildren === void 0 ? void 0 : arrayChildren.map(function (v) {
        if ( /*#__PURE__*/React.isValidElement(v)) {
          return _objectSpread(_objectSpread({}, v.props), {}, {
            label: v.props.label || v.props.children
          });
        }
        return {
          label: v
        };
      });
    }
    if (keys) {
      var _transformedOptions;
      transformedOptions = (_transformedOptions = transformedOptions) === null || _transformedOptions === void 0 ? void 0 : _transformedOptions.map(function (option) {
        return _objectSpread(_objectSpread({}, option), {}, {
          value: get_1(option, (keys === null || keys === void 0 ? void 0 : keys.value) || "value"),
          label: get_1(option, (keys === null || keys === void 0 ? void 0 : keys.label) || "label")
        });
      });
    }
    setCurrentOptions(transformedOptions);
    setTmpPropOptions(transformedOptions);
    setValueToOption(getValueToOption(children, options, keys) || {});
  }, [options, keys, children, reserveKeyword]);
  useEffect(function () {
    setSelectedOptions(function (oldSelectedOptions) {
      var valueKey = (keys === null || keys === void 0 ? void 0 : keys.value) || "value";
      var labelKey = (keys === null || keys === void 0 ? void 0 : keys.label) || "label";
      if (Array.isArray(value)) {
        return value.map(function (item) {
          if (valueType === "value") {
            return valueToOption[item] || oldSelectedOptions.find(function (option) {
              return get_1(option, valueKey) === item;
            }) || _defineProperty(_defineProperty({}, valueKey, item), labelKey, item);
          }
          return item;
        }).filter(Boolean);
      }
      if (value !== void 0 && value !== null) {
        if (valueType === "value") {
          return [valueToOption[value] || oldSelectedOptions.find(function (option) {
            return get_1(option, valueKey) === value;
          }) || _defineProperty(_defineProperty({}, valueKey, value), labelKey, value)].filter(Boolean);
        }
        return [value];
      }
      return [];
    });
  }, [value, keys, valueType, valueToOption, setSelectedOptions]);
  return {
    currentOptions: currentOptions,
    setCurrentOptions: setCurrentOptions,
    tmpPropOptions: tmpPropOptions,
    setTmpPropOptions: setTmpPropOptions,
    valueToOption: valueToOption,
    setValueToOption: setValueToOption,
    selectedOptions: selectedOptions,
    setSelectedOptions: setSelectedOptions
  };
}

export { UseOptions as default };
